### ISSUE TITLE
c:geo not starting after restore of data

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#10101]
While testing other issues I ran into a problem after this scenario:

- 2021.03.01-NB
- Go to device settings - Apps - c:geo and delete all data for c:geo
- Start c:geo (comes up with wizard)
- Complete the wizard (permissions, base folder, login to geocaching.com, skip advanced wizard)
- c:geo works normally
- Go to c:geo menu - Settings - Restore and select to restore caches and settings (used a backup done with yesterdays nightly, settings did not include account data)
- Backup is restored and restart is triggered by dialog

Result:
- Grey screen is shown, after some time splash screen logo shows up
- App runs into ANR
- Consecutive tries to start c:geo end up in ANR

Not yer knowing the exact cause of this problem, at least the following came to my mind:
What happens to previously existing SAF persistent folders in such case? The settings are restored, but the permissions might be lost...or is this also covered by the backup?

**Comment by eddiemuc** [comment]
> What happens to previously existing SAF persistent folders in such case? The settings are restored, but the permissions might be lost...or is this also covered by the backup?

That's a very good question. Haven't thought about this scenario yet.

AFAIK there is no way to "restore" permissions to folders in such a scenario, this is only possible via user interaction using folder selection intent. I would expect that if a folder setting is restored this way but points now to a folder w/o permission, that this is handled as if c:geo has lost permission in any other way (e.g. because user deleted or moved the folder on disk): error Toasts on Access and otherwise acting as if the folder is empty.

If c:geo does something else I would consider this a bug. Do you have any logs/errors for this?

**Comment by Lineflyer** [comment]
Not being able to start c:geo until now. Always ANR.
Could I place a log-properties.txt to try a log or does this not directly output the logfile....base dir folder should be present and permitted.

**Comment by eddiemuc** [comment]
No direct log output supported. The log properties just influence what is logged. I can build in an option to log into a file, but this requires a change.

Did you now really use your main c:geo installation?

**Comment by fm-sys** [comment]
I ran into a similar problem in the emulator. There is a stack trace, but no other SAF related logging in the logcat...

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: cgeo.geocaching, PID: 19449
    java.lang.ExceptionInInitializerError
        at cgeo.geocaching.storage.ContentStorage.get(ContentStorage.java:61)
        at cgeo.geocaching.utils.Log.<clinit>(Log.java:51)
        at cgeo.geocaching.utils.Log.setDebug(Log.java:87)
        at cgeo.geocaching.settings.Settings.<clinit>(Settings.java:118)
        at cgeo.geocaching.settings.Settings.getApplicationLocale(Settings.java:1535)
        at cgeo.geocaching.CgeoApplication.initApplicationLocale(CgeoApplication.java:116)
        at cgeo.geocaching.CgeoApplication.onCreate(CgeoApplication.java:52)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1182)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6460)
        at android.app.ActivityThread.access$1300(ActivityThread.java:219)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1859)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.net.Uri cgeo.geocaching.storage.ContentStorage.getUriForFolder(cgeo.geocaching.storage.Folder)' on a null object reference
        at cgeo.geocaching.utils.LocalizationUtils.getMultiPurposeString(LocalizationUtils.java:73)
        at cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:548)
        at cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:542)
        at cgeo.geocaching.storage.ContentStorage.getFolder(ContentStorage.java:505)
        at cgeo.geocaching.storage.ContentStorage.ensureFolder(ContentStorage.java:116)
        at cgeo.geocaching.storage.ContentStorage.reevaluateFolderDefaults(ContentStorage.java:110)
        at cgeo.geocaching.storage.ContentStorage.<init>(ContentStorage.java:102)
        at cgeo.geocaching.storage.ContentStorage.<clinit>(ContentStorage.java:58)
        at cgeo.geocaching.storage.ContentStorage.get(ContentStorage.java:61) 
        at cgeo.geocaching.utils.Log.<clinit>(Log.java:51) 
        at cgeo.geocaching.utils.Log.setDebug(Log.java:87) 
        at cgeo.geocaching.settings.Settings.<clinit>(Settings.java:118) 
        at cgeo.geocaching.settings.Settings.getApplicationLocale(Settings.java:1535) 
        at cgeo.geocaching.CgeoApplication.initApplicationLocale(CgeoApplication.java:116) 
        at cgeo.geocaching.CgeoApplication.onCreate(CgeoApplication.java:52) 
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1182) 
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6460) 
        at android.app.ActivityThread.access$1300(ActivityThread.java:219) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1859) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7356) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930) 
```

**Comment by MagpieFourtyTwo** [comment]
> What happens to previously existing SAF persistent folders in such case? The settings are restored, but the permissions might be lost...or is this also covered by the backup?

Permissions are not restored, for sure.
Stumbled upon this a few days ago, when I reverted to previous nightly because of the theme sync problem.
Nevertheless, I had no crash nor ANR. 
Only difference I see at first glance: I did not just clean data, but uninstalled c:geo completely ...
After installation and walking through the wizard, I restored data and settings (including account data), but had no ... hmmm ... themes? Don't know for sure what was missing, but s/th based in an SAF was not there ... so I went to setting, "changed" the setting of the appropriate SAF (to the already configured value), granted permission ... and then everything was fine again ...

**Comment by rsudev** [comment]
Pretty clear to see what the problem is: The ContentStorage-Constructor calls indirectly, in case of problems, a method, that expects the ContentStorage-singelton to be setup already....

**Comment by Lineflyer** [comment]
after several hours I checked again anf it meanwhile started.

Log is full of this error:
``
03-02 14:09:57.035 27292 27292 E AndroidRuntime: FATAL EXCEPTION: main
03-02 14:09:57.035 27292 27292 E AndroidRuntime: Process: cgeo.geocaching, PID: 27292
03-02 14:09:57.035 27292 27292 E AndroidRuntime: java.lang.ExceptionInInitializerError
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.get(ContentStorage.java:61)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.utils.Log.<clinit>(Log.java:51)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.utils.Log.setDebug(Log.java:87)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.settings.Settings.<clinit>(Settings.java:118)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.CgeoApplication.initApplicationLocale(CgeoApplication.java:116)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.CgeoApplication.onCreate(CgeoApplication.java:52)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1192)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.app.ActivityThread.handleMakeApplication(ActivityThread.java:7501)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:7440)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.app.ActivityThread.access$1500(ActivityThread.java:301)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2148)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:246)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8506)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.net.Uri cgeo.geocaching.storage.ContentStorage.getUriForFolder(cgeo.geocaching.storage.Folder)' on a null object reference
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.utils.LocalizationUtils.getMultiPurposeString(LocalizationUtils.java:73)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:548)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:542)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.getFolder(ContentStorage.java:505)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.ensureFolder(ContentStorage.java:116)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.reevaluateFolderDefaults(ContentStorage.java:110)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.<init>(ContentStorage.java:102)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.<clinit>(ContentStorage.java:58)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	... 17 more
03-02 14:10:28.489 27787 27787 E AndroidRuntime: FATAL EXCEPTION: main
03-02 14:10:28.489 27787 27787 E AndroidRuntime: Process: cgeo.geocaching, PID: 27787
03-02 14:10:28.489 27787 27787 E AndroidRuntime: java.lang.ExceptionInInitializerError
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.get(ContentStorage.java:61)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.utils.Log.<clinit>(Log.java:51)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.utils.Log.setDebug(Log.java:87)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.settings.Settings.<clinit>(Settings.java:118)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.CgeoApplication.initApplicationLocale(CgeoApplication.java:116)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.CgeoApplication.onCreate(CgeoApplication.java:52)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1192)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.app.ActivityThread.handleMakeApplication(ActivityThread.java:7501)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:7440)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.app.ActivityThread.access$1500(ActivityThread.java:301)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2148)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:246)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8506)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.net.Uri cgeo.geocaching.storage.ContentStorage.getUriForFolder(cgeo.geocaching.storage.Folder)' on a null object reference
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.utils.LocalizationUtils.getMultiPurposeString(LocalizationUtils.java:73)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:548)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:542)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.getFolder(ContentStorage.java:505)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.ensureFolder(ContentStorage.java:116)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.reevaluateFolderDefaults(ContentStorage.java:110)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.<init>(ContentStorage.java:102)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.<clinit>(ContentStorage.java:58)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	... 17 more
``

**Comment by eddiemuc** [comment]
hiopefully fixed with PR #10102.
This PR also adds the ability to write logs into a separate log file (configurable in log-properties.txt). Careful, this option is strictly experimental!

**Comment by Lineflyer** [comment]
Despite the fix of the crash, we should IMHO think about how to handle restore scenarios in combination with SAF.
Will open a new dedicated issue for this.
